### PR TITLE
[DNM] Concat KDF initial commit

### PR DESCRIPTION
**Pull request by rpiasetskyi**
Please do not merge

### CODE REVIEW

File: concat-kdf/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,81 @@
+#![no_std]
+
+use core::fmt;
+use digest::{generic_array::typenum::Unsigned, Digest, FixedOutputReset, Update};
+
+#[cfg(feature = "std")]
+extern crate std;
```
Comment (inorick) [inline concat-kdf/src/lib.rs:26]: Why is this needed?

File: concat-kdf/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+{
+    if secret.len() == 0 {
+        return Err(Error::NoSecret);
+    }
+
+    if key.len() == 0 {
+        return Err(Error::ZeroOutput);
+    }
+
+    if chunk_number(key.len(), D::OutputSize::USIZE) >= (u32::MAX as usize) {
```
Comment (inorick) [inline concat-kdf/src/lib.rs]: This can be replaced by precalcualting
`u32::MAX * D::OutputSize::USIZE`
and comparing `key.len()` against it.
You might need to use 64 bit data types during precalculation to avoid overflows.

File: concat-kdf/tests/tests.rs [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    );
+
+    assert_eq!(
+        concat_kdf::derive_key_into::<Sha512>(&[0u8; 42], &[], &mut [0u8; 0]),
+        Err(concat_kdf::Error::ZeroOutput)
+    );
+}
+
+#[test]
+#[ignore = "The test requires huge memory allocation."]
```
Comment (inorick) [inline concat-kdf/tests/tests.rs]: For testing purposes, it is possible to create a fake slice with the desired length using
https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html

File: concat-kdf/src/lib.rs (hunk 1 above)
Comment (rpiasetskyi) [reply concat-kdf/src/lib.rs:26]: https://doc.rust-lang.org/reference/names/preludes.html#the-no_std-attribute

```
error[E0433]: failed to resolve: use of undeclared crate or module `std`
  --> concat-kdf/src/lib.rs:74:13
   |
74 | ) -> Result<std::vec::Vec<u8>, Error>
   |             ^^^ use of undeclared crate or module `std`
```

`no_std` attribute prevents the `std` crate from being automatically added into scope.

File: concat-kdf/tests/tests.rs (hunk 3 above)
Comment (rpiasetskyi) [reply concat-kdf/tests/tests.rs]: Thanks!

File: concat-kdf/src/lib.rs (hunk 2 above)
Comment (rpiasetskyi) [reply concat-kdf/src/lib.rs]: Done.

File: concat-kdf/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    NoSecret,
+    /// The length of the output is zero.
+    NoOutput,
+    /// The length of the output is too big.
+    CounterOverflow,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        let msg = match self {
```
Comment (Nicoretti) [inline concat-kdf/src/lib.rs:41]: f.write_str( match self {
...
))

would also work and in some cases might read slightly better

File: concat-kdf/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+            Error::NoSecret => "Buffer for secret has zero length.",
+            Error::NoOutput => "Buffer for key has zero length.",
+            Error::CounterOverflow => "Requested key length is to big.",
+        };
+        f.write_str(msg)
+    }
+}
+
+#[cfg(feature = "std")]
```
Comment (Nicoretti) [inline concat-kdf/src/lib.rs:50]: depending on what you need from std, having alloc available may be enough. e.g. if only vec or box is needed.
It can be handy for porting the lib if only alloc is needed and not the entire std. Depends on the needs though.

https://doc.rust-lang.org/std/alloc/index.html

File: concat-kdf/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+    // Key length shall be less than or equal to hash output length * (2^32 - 1).
+    if OVERFLOW_IS_POSSIBLE && (key.len() >= D::OutputSize::USIZE * (u32::MAX as usize)) {
+        return Err(Error::CounterOverflow);
+    }
+
+    let mut digest = D::new();
+    let mut counter: u32 = 1;
+
+    for chunk in key.chunks_mut(D::OutputSize::USIZE) {
```
Comment (Nicoretti) [inline concat-kdf/src/lib.rs:82]: by adapting the iterator you could get rid of the counter (you may need a cast though)

e.g.

`for (index, chunk) in key.chunks_mut(...).enumerate()) { ...`

see also https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.enumerate

File: concat-kdf/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    Ok(())
+}
+
+/// Derives and returns `length` bytes key from `secret` and `other_info`.
+/// ```rust
+/// let key = concat_kdf::derive_key::<sha2::Sha256>(b"top-secret", b"info", 42).unwrap();
+/// ```
+#[cfg(feature = "std")]
```
Comment (Nicoretti) [inline concat-kdf/src/lib.rs:97]: also consider restricting it to `alloc` if thats reasonable

File: concat-kdf/tests/tests.rs [hunk 8]
```diff
@@ -0,0 +1,200 @@
+use digest::{Digest, FixedOutputReset};
+use hex_literal::hex;
+use sha2::{Sha224, Sha256, Sha512};
+
+struct Test<'a> {
```
Comment (Nicoretti) [inline concat-kdf/tests/tests.rs:5]: TestData? Fixture?  this isn't a test isn't it?

File: concat-kdf/tests/tests.rs [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+    secret: &'a [u8],
+    other_info: &'a [u8],
+    expected_key: &'a [u8],
+}
+
+fn test_input_output<D>(tests: &[Test])
```
Comment (Nicoretti) [inline concat-kdf/tests/tests.rs:11]: maybe something like `test_key_derivation(fixture: Fixture)` and then call it for all fixtures via. macro or in some sort of test like `test_all_...`

File: concat-kdf/tests/tests.rs [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+{
+    for Test {
+        secret,
+        other_info,
+        expected_key,
+    } in tests.iter()
+    {
+        for key_length in 1..expected_key.len() {
+            let mut key = vec![0u8; key_length];
+            assert!(concat_kdf::derive_key_into::<D>(secret, other_info, &mut key).is_ok());
```
Comment (Nicoretti) [inline concat-kdf/tests/tests.rs:23]: maybe create custom assert/macro with a speaking name e.g. `assert_key_derivation(...)`
which does all the needed checks (the 2 asserts you are currently using)

File: concat-kdf/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+#[cfg_attr(docsrs, doc(cfg(feature = "std")))]
+impl ::std::error::Error for Error {}
+
+/// Derives `key` in-place from `secret` and `other_info`.
+/// ```rust
+/// let mut key = [0u8; 42];
+/// concat_kdf::derive_key_into::<sha2::Sha256>(b"top-secret", b"info", &mut key).unwrap();
+/// ```
+pub fn derive_key_into<D>(secret: &[u8], other_info: &[u8], key: &mut [u8]) -> Result<(), Error>
```
Comment (Nicoretti) [inline concat-kdf/src/lib.rs:59]: other info == salt?

File: concat-kdf/src/lib.rs (hunk 6 above)
Comment (rpiasetskyi) [reply concat-kdf/src/lib.rs:82]: Unfortunately, `counter` has to start from `1` and `enumerate()` starts from `0` (additional `+1` is needed and cast).

File: concat-kdf/src/lib.rs (hunk 11 above)
Comment (rpiasetskyi) [reply concat-kdf/src/lib.rs:59]: It can be called like this but in the [standard (section 5.8.1)](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-56ar.pdf) `OtherInfo` is used.

File: concat-kdf/tests/tests.rs (hunk 8 above)
Comment (rpiasetskyi) [reply concat-kdf/tests/tests.rs:5]: Done. Thanks!

File: concat-kdf/src/lib.rs (hunk 4 above)
Comment (rpiasetskyi) [reply concat-kdf/src/lib.rs:41]: Done. Thanks!
